### ISSUE TITLE
subprocess server memory leak

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#1930]
I noticed when working on #1922 that valgrind would find memory leaks in the broker exec service if I didn't send it a `cmb.disconnect` message on teardown.

As discussed in #1926, I probably should be taking steps to ensure subprocesses are complete before I destroy their `flux_subprocess_t` down in [worker.c](https://github.com/flux-framework/flux-core/blob/372d41b58791001dd4778beafeee4f67e37ecdf9/src/modules/job-ingest/worker.c#L307), but it seemed like the server side should still probably try not to leak.

I'm finding a simpler reproducer elusive at the moment, so here's the long form reproducer as a placeholder:  make this change and run the valgrind sharness test:

```diff
diff --git a/src/modules/job-ingest/validate.c b/src/modules/job-ingest/validate.c
index bc2dcf51..5c37fc47 100644
--- a/src/modules/job-ingest/validate.c
+++ b/src/modules/job-ingest/validate.c
@@ -62,7 +62,7 @@ struct validate {
 
 /* Trigger cleanpup in broker exec service.
  */
-static void disconnect (flux_t *h)
+void disconnect (flux_t *h)
 {
     flux_future_t *f;
 
@@ -78,7 +78,7 @@ void validate_destroy (struct validate *v)
         int i;
         for (i = 0; i < MAX_WORKER_COUNT; i++)
             worker_destroy (v->worker[i]);
-        disconnect (v->h);
+        //disconnect (v->h);
         free (v);
         errno = saved_errno;
     }
diff --git a/t/valgrind/workload.d/wreck b/t/valgrind/workload.d/wreck
index 5c158259..5d698272 100755
--- a/t/valgrind/workload.d/wreck
+++ b/t/valgrind/workload.d/wreck
@@ -1,5 +1,5 @@
 #!/bin/bash
-
+exit
 size=$(flux getattr size)
 NJOBS=10
 
```

**Comment by garlick** [comment]
See [valgrind.log](https://github.com/flux-framework/flux-core/files/2771117/valgrind.log)

**Comment by chu11** [comment]
still no luck reproducing.

**Comment by garlick** [comment]
I just reproduced on my home system on current master.  There is some timing involved though, because it doesn't reproduce if `wreck` runs after `job` (which is why I commented it out in the patch).  I'll try on catalyst when I get in.

**Comment by grondo** [comment]
Extending the `worker_activity_timeout` would probably also help the reproducer. The wreck test takes a long time, and probably gives the workers time to be terminated due to inactivity. (Sorry if that's obvious!)

You could probably reproduce with the `cron` module (with a sleep 30 job), but I doubt the `cron` module cleans up at exit either, so you might get more leaks than you care to see. :disappointed:

**Comment by chu11** [comment]
I doubled the ```worker_inactivity_timeout``` in the ingest module and that seems to have done the trick!  Looks to be about what @garlick saw.  Thanks @grondo.

**Comment by chu11** [comment]
So I think I got the basics of the fix.  Basically we have a process on the client side that has given notice it wants to terminate (via close STDIN) and has destroyed its ```flux_subprocess_t *```.  However, the remote side had not finished cleaning up the process when the server was brought down.  The server going down did not properly cleanup any lingering subprocesses that had not yet completed.

I think the ```cmb.disconnect``` that was done in PR #1922 increased the probability of the subprocess cleaning up properly for the server brought down, but it doesn't guarantee it.

Still trying to figure out the best way to fix this, as there's (to my knowledge) no safe way to iterate through a ```zhash_t``` and ```zhash_delete``` all the items within it.  So I think a refactoring to use a free function is what I got to do.
